// sql-formatter/lib/lexer/Tokenizer.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _token = require("./token");

var regex = _interopRequireWildcard(require("./regexFactory"));

var _TokenizerEngine = _interopRequireDefault(require("./TokenizerEngine"));

var _regexUtil = require("./regexUtil");

var _utils = require("../utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Tokenizer = /*#__PURE__*/function () {
  function Tokenizer(cfg) {
    _classCallCheck(this, Tokenizer);

    this.cfg = cfg;

    _defineProperty(this, "dialectRules", void 0);

    this.dialectRules = this.buildDialectRules(cfg);
  }

  _createClass(Tokenizer, [{
    key: "tokenize",
    value: function tokenize(input, paramTypesOverrides) {
      var rules = _objectSpread(_objectSpread({}, this.dialectRules), this.buildParamRules(this.cfg, paramTypesOverrides));

      var tokens = new _TokenizerEngine["default"](rules).tokenize(input);
      return this.cfg.postProcess ? this.cfg.postProcess(tokens) : tokens;
    } // These rules can be cached as they only depend on
    // the Tokenizer config options specified for each SQL dialect

  }, {
    key: "buildDialectRules",
    value: function buildDialectRules(cfg) {
      var _cfg$lineCommentTypes, _cfg$reservedPhrases, _cfg$operators, _this$validRules;

      return this.validRules((_this$validRules = {}, _defineProperty(_this$validRules, _token.TokenType.BLOCK_COMMENT, {
        regex: new RegExp("(\\/\\*(?:(?![])[\\s\\S])*?(?:\\*\\/|$))", "y")
      }), _defineProperty(_this$validRules, _token.TokenType.LINE_COMMENT, {
        regex: regex.lineComment((_cfg$lineCommentTypes = cfg.lineCommentTypes) !== null && _cfg$lineCommentTypes !== void 0 ? _cfg$lineCommentTypes : ['--'])
      }), _defineProperty(_this$validRules, _token.TokenType.QUOTED_IDENTIFIER, {
        regex: regex.string(cfg.identTypes)
      }), _defineProperty(_this$validRules, _token.TokenType.NUMBER, {
        regex: new RegExp("(?:0x[0-9A-Fa-f]+|0b[01]+|(?:\\x2D[\\t-\\r \\xA0\\u1680\\u2000-\\u200A\\u2028\\u2029\\u202F\\u205F\\u3000\\uFEFF]*)?[0-9]+(?:\\.[0-9]*)?(?:[Ee][\\+\\x2D]?[0-9]+(?:\\.[0-9]+)?)?)(?![0-9A-Z_a-z])", "y")
      }), _defineProperty(_this$validRules, _token.TokenType.CASE, {
        regex: new RegExp("CA[S\\u017F]E\\b", "iy"),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.END, {
        regex: new RegExp("END\\b", "iy"),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.BETWEEN, {
        regex: new RegExp("BETWEEN\\b", "iy"),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.LIMIT, {
        regex: cfg.reservedCommands.includes('LIMIT') ? new RegExp("LIMIT\\b", "iy") : undefined,
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.RESERVED_COMMAND, {
        regex: regex.reservedWord(cfg.reservedCommands, cfg.identChars),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.RESERVED_SELECT, {
        regex: regex.reservedWord(cfg.reservedSelect, cfg.identChars),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.RESERVED_SET_OPERATION, {
        regex: regex.reservedWord(cfg.reservedSetOperations, cfg.identChars),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.RESERVED_DEPENDENT_CLAUSE, {
        regex: regex.reservedWord(cfg.reservedDependentClauses, cfg.identChars),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.RESERVED_JOIN, {
        regex: regex.reservedWord(cfg.reservedJoins, cfg.identChars),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.RESERVED_PHRASE, {
        regex: regex.reservedWord((_cfg$reservedPhrases = cfg.reservedPhrases) !== null && _cfg$reservedPhrases !== void 0 ? _cfg$reservedPhrases : [], cfg.identChars),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.AND, {
        regex: new RegExp("AND\\b", "iy"),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.OR, {
        regex: new RegExp("OR\\b", "iy"),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.XOR, {
        regex: cfg.supportsXor ? new RegExp("XOR\\b", "iy") : undefined,
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.RESERVED_FUNCTION_NAME, {
        regex: regex.reservedWord(cfg.reservedFunctionNames, cfg.identChars),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.RESERVED_KEYWORD, {
        regex: regex.reservedWord(cfg.reservedKeywords, cfg.identChars),
        value: toCanonical
      }), _defineProperty(_this$validRules, _token.TokenType.VARIABLE, {
        regex: cfg.variableTypes ? regex.variable(cfg.variableTypes) : undefined
      }), _defineProperty(_this$validRules, _token.TokenType.STRING, {
        regex: regex.string(cfg.stringTypes)
      }), _defineProperty(_this$validRules, _token.TokenType.IDENTIFIER, {
        regex: regex.identifier(cfg.identChars)
      }), _defineProperty(_this$validRules, _token.TokenType.DELIMITER, {
        regex: new RegExp(";", "y")
      }), _defineProperty(_this$validRules, _token.TokenType.COMMA, {
        regex: new RegExp("[,]", "y")
      }), _defineProperty(_this$validRules, _token.TokenType.OPEN_PAREN, {
        regex: regex.parenthesis('open', cfg.extraParens)
      }), _defineProperty(_this$validRules, _token.TokenType.CLOSE_PAREN, {
        regex: regex.parenthesis('close', cfg.extraParens)
      }), _defineProperty(_this$validRules, _token.TokenType.OPERATOR, {
        regex: regex.operator('+-/%&|^><=.:$@#?~!', ['<>', '<=', '>=', '!='].concat(_toConsumableArray((_cfg$operators = cfg.operators) !== null && _cfg$operators !== void 0 ? _cfg$operators : [])))
      }), _defineProperty(_this$validRules, _token.TokenType.ASTERISK, {
        regex: new RegExp("\\*", "y")
      }), _this$validRules));
    } // These rules can't be blindly cached as the paramTypesOverrides object
    // can differ on each invocation of the format() function.

  }, {
    key: "buildParamRules",
    value: function buildParamRules(cfg, paramTypesOverrides) {
      var _cfg$paramTypes, _cfg$paramTypes2, _cfg$paramTypes3, _cfg$paramTypes4, _this$validRules2;

      // Each dialect has its own default parameter types (if any),
      // but these can be overriden by the user of the library.
      var paramTypes = {
        named: (paramTypesOverrides === null || paramTypesOverrides === void 0 ? void 0 : paramTypesOverrides.named) || ((_cfg$paramTypes = cfg.paramTypes) === null || _cfg$paramTypes === void 0 ? void 0 : _cfg$paramTypes.named) || [],
        quoted: (paramTypesOverrides === null || paramTypesOverrides === void 0 ? void 0 : paramTypesOverrides.quoted) || ((_cfg$paramTypes2 = cfg.paramTypes) === null || _cfg$paramTypes2 === void 0 ? void 0 : _cfg$paramTypes2.quoted) || [],
        numbered: (paramTypesOverrides === null || paramTypesOverrides === void 0 ? void 0 : paramTypesOverrides.numbered) || ((_cfg$paramTypes3 = cfg.paramTypes) === null || _cfg$paramTypes3 === void 0 ? void 0 : _cfg$paramTypes3.numbered) || [],
        positional: typeof (paramTypesOverrides === null || paramTypesOverrides === void 0 ? void 0 : paramTypesOverrides.positional) === 'boolean' ? paramTypesOverrides.positional : (_cfg$paramTypes4 = cfg.paramTypes) === null || _cfg$paramTypes4 === void 0 ? void 0 : _cfg$paramTypes4.positional
      };
      return this.validRules((_this$validRules2 = {}, _defineProperty(_this$validRules2, _token.TokenType.NAMED_PARAMETER, {
        regex: regex.parameter(paramTypes.named, regex.identifierPattern(cfg.paramChars || cfg.identChars)),
        key: function key(v) {
          return v.slice(1);
        }
      }), _defineProperty(_this$validRules2, _token.TokenType.QUOTED_PARAMETER, {
        regex: regex.parameter(paramTypes.quoted, regex.stringPattern(cfg.identTypes)),
        key: function key(v) {
          return function (_ref) {
            var tokenKey = _ref.tokenKey,
                quoteChar = _ref.quoteChar;
            return tokenKey.replace(new RegExp((0, _regexUtil.escapeRegExp)('\\' + quoteChar), 'gu'), quoteChar);
          }({
            tokenKey: v.slice(2, -1),
            quoteChar: v.slice(-1)
          });
        }
      }), _defineProperty(_this$validRules2, _token.TokenType.NUMBERED_PARAMETER, {
        regex: regex.parameter(paramTypes.numbered, '[0-9]+'),
        key: function key(v) {
          return v.slice(1);
        }
      }), _defineProperty(_this$validRules2, _token.TokenType.POSITIONAL_PARAMETER, {
        regex: paramTypes.positional ? new RegExp("[?]", "y") : undefined
      }), _this$validRules2));
    } // filters out rules for token types whose regex is undefined

  }, {
    key: "validRules",
    value: function validRules(rules) {
      return Object.fromEntries(Object.entries(rules).filter(function (_ref2) {
        var _ref3 = _slicedToArray(_ref2, 2),
            _ = _ref3[0],
            rule = _ref3[1];

        return rule.regex;
      }));
    }
  }]);

  return Tokenizer;
}();
/**
 * Converts keywords (and keyword sequences) to canonical form:
 * - in uppercase
 * - single spaces between words
 */


exports["default"] = Tokenizer;

var toCanonical = function toCanonical(v) {
  return (0, _utils.equalizeWhitespace)(v.toUpperCase());
};

module.exports = exports.default;
