// sql-formatter/lib/lexer/token.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.testToken = exports.isToken = exports.isReserved = exports.isParameter = exports.isLogicalOperator = exports.TokenType = exports.EOF_TOKEN = void 0;

/** Token type enum for all possible Token categories */
var TokenType;
/** Struct to store the most basic cohesive unit of language grammar */

exports.TokenType = TokenType;

(function (TokenType) {
  TokenType["QUOTED_IDENTIFIER"] = "QUOTED_IDENTIFIER";
  TokenType["IDENTIFIER"] = "IDENTIFIER";
  TokenType["STRING"] = "STRING";
  TokenType["VARIABLE"] = "VARIABLE";
  TokenType["RESERVED_KEYWORD"] = "RESERVED_KEYWORD";
  TokenType["RESERVED_FUNCTION_NAME"] = "RESERVED_FUNCTION_NAME";
  TokenType["RESERVED_PHRASE"] = "RESERVED_PHRASE";
  TokenType["RESERVED_DEPENDENT_CLAUSE"] = "RESERVED_DEPENDENT_CLAUSE";
  TokenType["RESERVED_SET_OPERATION"] = "RESERVED_SET_OPERATION";
  TokenType["RESERVED_COMMAND"] = "RESERVED_COMMAND";
  TokenType["RESERVED_SELECT"] = "RESERVED_SELECT";
  TokenType["RESERVED_JOIN"] = "RESERVED_JOIN";
  TokenType["ARRAY_IDENTIFIER"] = "ARRAY_IDENTIFIER";
  TokenType["ARRAY_KEYWORD"] = "ARRAY_KEYWORD";
  TokenType["CASE"] = "CASE";
  TokenType["END"] = "END";
  TokenType["LIMIT"] = "LIMIT";
  TokenType["BETWEEN"] = "BETWEEN";
  TokenType["AND"] = "AND";
  TokenType["OR"] = "OR";
  TokenType["XOR"] = "XOR";
  TokenType["OPERATOR"] = "OPERATOR";
  TokenType["COMMA"] = "COMMA";
  TokenType["ASTERISK"] = "ASTERISK";
  TokenType["OPEN_PAREN"] = "OPEN_PAREN";
  TokenType["CLOSE_PAREN"] = "CLOSE_PAREN";
  TokenType["LINE_COMMENT"] = "LINE_COMMENT";
  TokenType["BLOCK_COMMENT"] = "BLOCK_COMMENT";
  TokenType["NUMBER"] = "NUMBER";
  TokenType["NAMED_PARAMETER"] = "NAMED_PARAMETER";
  TokenType["QUOTED_PARAMETER"] = "QUOTED_PARAMETER";
  TokenType["NUMBERED_PARAMETER"] = "NUMBERED_PARAMETER";
  TokenType["POSITIONAL_PARAMETER"] = "POSITIONAL_PARAMETER";
  TokenType["DELIMITER"] = "DELIMITER";
  TokenType["EOF"] = "EOF";
})(TokenType || (exports.TokenType = TokenType = {}));

/**
 * For use as a "missing token"
 * e.g. in lookAhead and lookBehind to avoid dealing with null values
 */
var EOF_TOKEN = {
  type: TokenType.EOF,
  raw: '«EOF»',
  text: '«EOF»',
  start: Infinity,
  end: Infinity
};
/** Checks if two tokens are equivalent */

exports.EOF_TOKEN = EOF_TOKEN;

var testToken = function testToken(compareToken) {
  return function (token) {
    return token.type === compareToken.type && token.text === compareToken.text;
  };
};
/** Util object that allows for easy checking of Reserved Keywords */


exports.testToken = testToken;
var isToken = {
  ARRAY: testToken({
    text: 'ARRAY',
    type: TokenType.RESERVED_KEYWORD
  }),
  BY: testToken({
    text: 'BY',
    type: TokenType.RESERVED_KEYWORD
  }),
  SET: testToken({
    text: 'SET',
    type: TokenType.RESERVED_COMMAND
  }),
  STRUCT: testToken({
    text: 'STRUCT',
    type: TokenType.RESERVED_KEYWORD
  }),
  WINDOW: testToken({
    text: 'WINDOW',
    type: TokenType.RESERVED_COMMAND
  })
};
/** Checks if token is any Reserved Keyword or Command */

exports.isToken = isToken;

var isReserved = function isReserved(token) {
  return token.type === TokenType.RESERVED_KEYWORD || token.type === TokenType.RESERVED_FUNCTION_NAME || token.type === TokenType.RESERVED_PHRASE || token.type === TokenType.RESERVED_DEPENDENT_CLAUSE || token.type === TokenType.RESERVED_COMMAND || token.type === TokenType.RESERVED_SELECT || token.type === TokenType.RESERVED_SET_OPERATION || token.type === TokenType.RESERVED_JOIN || token.type === TokenType.ARRAY_KEYWORD || token.type === TokenType.CASE || token.type === TokenType.END || token.type === TokenType.LIMIT || token.type === TokenType.BETWEEN || token.type === TokenType.AND || token.type === TokenType.OR || token.type === TokenType.XOR;
};
/** checks if token is one of the parameter tokens */


exports.isReserved = isReserved;

var isParameter = function isParameter(token) {
  return token.type === TokenType.NUMBERED_PARAMETER || token.type === TokenType.NAMED_PARAMETER || token.type === TokenType.POSITIONAL_PARAMETER || token.type === TokenType.QUOTED_PARAMETER;
};

exports.isParameter = isParameter;

var isLogicalOperator = function isLogicalOperator(token) {
  return token.type === TokenType.AND || token.type === TokenType.OR || token.type === TokenType.XOR;
};

exports.isLogicalOperator = isLogicalOperator;
